### DISCUSSION TITLE
working-with-the-result-pattern-part-3-adding-more-extensions/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#439] [Announcements]
# working-with-the-result-pattern-part-3-adding-more-extensions/

In this post I show how to extend your LINQ-based result pattern code to handle more scenarios, such as Task<Result<T>> and collections of Result<T>

https://andrewlock.net/working-with-the-result-pattern-part-3-adding-more-extensions/

**Comment by Plevi1337** [comment]
The "monadic and applicative styles" link is a 404

**Comment by altinoren** [comment]
A couple years ago I've mentioned `Result<T>` in a comment on HN and it sparked some thoughtful comments about states and use in other languages. Leaving it here as an additional discussion point.

https://news.ycombinator.com/item?id=23543497

**Comment by Evolva** [comment]
In our case we decide to implement 4 versions of SelectMany

```cs
//Sync then Sync
public static Result<TResult> SelectMany<TSource, TMiddle, TResult>(
        this Result<TSource> source,
        Func<TSource, Result<TMiddle>> collectionSelector,
        Func<TSource, TMiddle, TResult> resultSelector)

//Sync then Async
public static async Task<Result<TResult>> SelectMany<TSource, TMiddle, TResult>(
        this Result<TSource> source,
        Func<TSource, Task<Result<TMiddle>>> collectionSelector,
        Func<TSource, TMiddle, TResult> resultSelector)

//Async then Sync
public static async Task<Result<TResult>> SelectMany<TSource, TMiddle, TResult>(
        this Task<Result<TSource>> source,
        Func<TSource, Result<TMiddle>> collectionSelector,
        Func<TSource, TMiddle, TResult> resultSelector)

//Async then Async
public static async Task<Result<TResult>> SelectMany<TSource, TMiddle, TResult>(
        this Task<Result<TSource>> source,
        Func<TSource, Task<Result<TMiddle>>> collectionSelector,
        Func<TSource, TMiddle, TResult> resultSelector)
```

This way we are able to mix and match functions that return `Result<TResult>` and `Task<Result<TResult>>` and avoid the `Task.FromResult(...)` in LINQ statements

**Comment by VictorioBerra** [comment]
I have a question that I struggle with regarding the results pattern. If I have a method in service A that calls multiple methods in service B, what result should service A return?

I found it difficult to have each layer of my app return Result types and I'm still not sure the appropriate approach.

**Comment by repodn** [comment]
I get it, but it's not easy. If I don't see the code for 14 days and then sit down again, it will take me a while to understand it again.

**Comment by repodn** [comment]
Looking at Aggregate: when the second item out of five fails, three iterations are executed unnecessarily?

**Comment by alexandersorokin** [comment]
Thanks, Andrew, for the great article.

Handling `Result<T>` and `Task<Result<T>>` with do-notation is not enough in real applications. We should handle `Task<T>` as well. And this is very hard. I have seen one implementation that tries to handle it, but it fails if there are three or more `from in` statements.

To handle all cases correctly, I had to generate several hundred SelectMany method overloads. And I had to check the C# specification to split methods by namespaces correctly. And it becomes more complex if we add ValueTask and other features.

You can check a correct implantation of do-notation here: https://github.com/skbkontur/results?tab=readme-ov-file#do-notation.

There are [hundreds of tests](https://github.com/skbkontur/results/blob/master/Kontur.Tests.Results/Conversion/Combinations/Linq/Result/TValueTValue/SelectMany/Results4/Task_Should.cs#L57) that check that it works correctly. Other open source result implementations I found don't support ValueTask, Task<T> (without Result) or having more than three `from in` statements.

**Comment by seungyongshim** [comment]
good another monad tutorial
